internal/users/db/testutils.go

```diff
@@ -4,149 +4,209 @@ package db
-				}
-				if err := bucket.Put([]byte(key), []byte(val)); err != nil {
-					panic("programming error: put called in a RO transaction")
-				}
-			}
-		}
-		return nil
-	})
+	// Now dump the SQLite database to YAML
+	s := Z_ForTests_DumpNormalizedYAML(t, db)
+
+	// Store the YAML in a new db file
+	err = os.WriteFile(src, []byte(s), 0600)
+	require.NoError(t, err)
 }
 
-// Z_ForTests_CreateDBFromYAML creates the database inside destDir and loads the src file content into it.
+// Z_ForTests_CreateDBFromYAML creates the bbolt database inside destDir and loads the src file content into it.
 //
 // nolint:revive,nolintlint // We want to use underscores in the function name here.
-func Z_ForTests_CreateDBFromYAML(t *testing.T, src, destDir string) {
+func Z_ForTests_CreateDBFromYAMLReader(t *testing.T, r io.Reader, destDir string) {
 	t.Helper()
-	testsdetection.MustBeTesting()
+	createDBFromYAMLReader(t, r, destDir)
+}
 
-	if !path.IsAbs(src) {
-		wd, err := os.Getwd()
-		require.NoError(t, err, "Setup: should be able to get working directory")
-		src = filepath.Join(wd, src)
-	}
+func createDBFromYAML(t *testing.T, src, destDir string) {
+	t.Helper()
+	log.Debugf(context.Background(), "Loading SQLite database from %s", src)
 
 	f, err := os.Open(src)
-	require.NoError(t, err, "Setup: should be able to read source file")
-	defer f.Close()
+	require.NoError(t, err)
 
-	err = Z_ForTests_FromYAML(f, destDir)
-	require.NoError(t, err, "Setup: should be able to write database file")
+	createDBFromYAMLReader(t, f, destDir)
 }
 
-// Z_ForTests_DBName returns the name of the database.
-//
-// nolint:revive,nolintlint // We want to use underscores in the function name here.
-func Z_ForTests_DBName() string {
-	testsdetection.MustBeTesting()
-	return dbName
+func createDBFromYAMLReader(t *testing.T, r io.Reader, destDir string) {
+	t.Helper()
+
+	yamlData, err := io.ReadAll(r)
+	require.NoError(t, err)
+
+	// unmarshal the content into a map.
+	dbContent := make(map[string][]map[string]string)
+	err = yaml.Unmarshal(yamlData, dbContent)
+	require.NoError(t, err)
+
+	db, err := New(destDir)
+	require.NoError(t, err)
+	defer func() {
+		err := db.Close()
+		require.NoError(t, err)
+	}()
+
+	tablesInOrder := []string{"users", "groups", "users_to_groups"}
+
+	// Insert data
+	for _, table := range tablesInOrder {
+		records, exists := dbContent[table]
+		if !exists {
+			continue
+		}
+
+		for _, record := range records {
+			columns := ""
+			values := ""
+			var vals []interface{}
+
+			for col, val := range record {
+				if columns != "" {
+					columns += ", "
+					values += ", "
+				}
+				columns += col
+				values += "?"
+				vals = append(vals, val)
+			}
+			log.Debugf(context.Background(), "Inserting into %s: %s", table, vals)
+
+			//nolint:gosec // We don't care about SQL injection in our tests.
+			query := fmt.Sprintf("INSERT INTO %s (%s) VALUES (%s)", table, columns, values)
```

Comment:
Is there any potential string escaping that we may miss here?

Reply:
None that I know of. Is it worth spending time trying to find one? I don't see why we should care about the SQL injection warning here.

Reply:
Well, not sure if a malformed username query could lead to do something... I'd still like to block if possible (or ignore me if the user name is already sanitized enough).

Reply:
I don't understand what you mean. Which username query? This is an INSERT statement in our testutils, which is used to insert testdata into a test database. I don't see why we should care about SQL injection in that context.

Reply:
Oh I think I had linked an invalid line, but my comment was generic to ensure we didn't end up doing unwanted queries because of non-sanitized data.